test/python/pulse/test_builder.py

```diff
@@ -0,0 +1,969 @@
+                pulse.delay(3, d0)
+                with pulse.align_left():
+                    pulse.delay(5, d1)
+                    pulse.delay(7, d0)
+
+        reference = pulse.Schedule()
+        # d0
+        reference += instructions.Delay(3, d0)
+        reference += instructions.Delay(7, d0)
+        # d1
+        reference = reference.insert(3, instructions.Delay(5, d1))
+        # d2
+        reference += instructions.Delay(11, d2)
+        self.assertEqual(schedule, reference)
+
+    def test_align_right(self):
+        """Test the right alignment context."""
+        d0 = pulse.DriveChannel(0)
+        d1 = pulse.DriveChannel(1)
+        d2 = pulse.DriveChannel(2)
+
+        with pulse.build() as schedule:
+            with pulse.align_right():
+                pulse.delay(11, d2)
+                pulse.delay(3, d0)
+                with pulse.align_right():
+                    pulse.delay(5, d1)
+                    pulse.delay(7, d0)
+
+        reference = pulse.Schedule()
+        # d0
+        reference = reference.insert(1, instructions.Delay(3, d0))
+        reference = reference.insert(4, instructions.Delay(7, d0))
+        # d1
+        reference = reference.insert(6, instructions.Delay(5, d1))
+        # d2
+        reference += instructions.Delay(11, d2)
+
+        self.assertEqual(schedule, reference)
+
+    def test_group(self):
+        """Test the grouping context."""
+        d0 = pulse.DriveChannel(0)
+        d1 = pulse.DriveChannel(1)
+
+        with pulse.build() as schedule:
+            pulse.delay(3, d0)
+            with pulse.group():
+                pulse.delay(5, d1)
+                pulse.delay(7, d0)
+
+        reference = pulse.Schedule()
+        # d0
+        reference += instructions.Delay(3, d0)
+        reference += instructions.Delay(7, d0)
+        # d1
+        reference = reference.insert(3, instructions.Delay(5, d1))
+        self.assertEqual(schedule, reference)
+
+    def test_inline(self):
+        """Test the inlining context."""
+        d0 = pulse.DriveChannel(0)
+        d1 = pulse.DriveChannel(1)
+
+        with pulse.build() as schedule:
+            pulse.delay(3, d0)
+            with pulse.inline():
+                pulse.delay(5, d1)
+                pulse.delay(7, d0)
+
+        reference = pulse.Schedule()
+        # d0
+        reference += instructions.Delay(3, d0)
+        reference += instructions.Delay(7, d0)
+        # d1
+        reference += instructions.Delay(5, d1)
+
+        self.assertEqual(schedule, reference)
+
+    def test_transpiler_settings(self):
+        """Test the transpiler settings context.
+
+        Tests that two cx gates are optimized away with higher optimization level.
+        """
+        twice_cx_qc = circuit.QuantumCircuit(2)
+        twice_cx_qc.cx(0, 1)
+        twice_cx_qc.cx(0, 1)
+
+        with pulse.build(self.backend) as schedule:
+            with pulse.transpiler_settings(optimization_level=0):
```

Comment:
This is something that will be deprecated with the circuit builder, right?

Reply:
Yes, although similar functionality might be required in the circuit builder @nkanazawa1989 and Kentaro Heya (his intern) have found it useful.